onnxscript/rewriter/llama_rule_sets.py

```diff
@@ -2,13 +2,73 @@
 # Licensed under the MIT License.
 from __future__ import annotations
 
+import numpy as np
+import onnx.numpy_helper
+
 import onnxscript.ir as ir
 import onnxscript.rewriter.no_op as no_op
 import onnxscript.rewriter.pattern as orp
 
 op = orp.onnxop
 
 
+class CastIdentity(orp.RewriteRuleAsClass):
+    """Replaces ``Cast(., to=to)`` by ``Identity`` if possible."""
+
+    @classmethod
+    def pattern(cls, op, x, to: int):
+        return op.Cast(x, to=to)
+
+    @classmethod
+    def rewrite(cls, op, x: ir.Value, to: int):
+        return op.Identity(x)
+
+    @classmethod
+    def check(cls, context, x, to: int) -> bool:
+        return x.dtype == to.value
+
+
+class CastCast(orp.RewriteRuleAsClass):
+    """Replaces ``Cast(Cast(X, ...), to=to)`` by ``Cast(X, to=to)``."""
+
+    @classmethod
+    def pattern(cls, op, x, to: int, to0: int):
+        return op.Cast(op.Cast(x, to=to0), to=to)
+
+    @classmethod
+    def rewrite(cls, op, x: ir.Value, to: int, to0: int):
+        return op.Cast(x, to=to)
+
+
+class ExpandIdentity(orp.RewriteRuleAsClass):
+    """Replaces ``Expand(., shape)`` by ``Identity`` if possible."""
+
+    @classmethod
+    def pattern(cls, op, x, shape):
+        return op.Expand(x, shape)
+
+    @classmethod
+    def rewrite(cls, op, x: ir.Value, shape: ir.Value):
+        return op.Identity(x)
+
+    @classmethod
+    def check(cls, context, x, shape) -> bool:
+        shape_x = x.shape
+        return shape_x.dims == tuple(shape.const_value.numpy().tolist())
+
+
+class ReshapeReshape(orp.RewriteRuleAsClass):
+    """Replaces ``Reshape(Reshape(X, ...), shape)`` by ``Reshape(X, shape)``."""
```

Comment:
I think this is not always valid ... specifically, when -1 and 0 are is used in the shape. For example, consider `Reshape(Reshape(X, [128, 128, 128]), [256, 0, -1]))` Would checking that there are no 0 and -1 in shape work?

Reply:
Right, I forgot that. I'll add those verification.